infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java

```diff
@@ -54,83 +64,139 @@ public LogEntryWriter(CorfuRuntime rt, LogReplicationConfig config, LogReplicati
+
+                        // Validate the message metadata with the local metadata table
+                        // TODO: This validation should be performed only for the topology config id after the
+                        //  metadata table's schema is changed
+                        if (topologyConfigId != persistedTopologyConfigId || baseSnapshotTs != persistedSnapshotStart ||
+                            baseSnapshotTs != persistedSnapshotDone || prevTs != persistedBatchTs) {
+                            log.warn("Message metadata mismatch. Skip applying message {}, persistedTopologyConfigId={}," +
+                                    "persistedSnapshotStart={}, persistedSnapshotDone={}, persistedBatchTs={}",
+                                txMessage.getMetadata(), persistedTopologyConfigId, persistedSnapshotStart,
+                                persistedSnapshotDone, persistedBatchTs);
+                            throw new IllegalArgumentException("Cannot apply log entry message due to metadata mismatch");
+                        }
+
+                        // Skip Opaque entries with timestamp that are not larger than persistedOpaqueEntryTs
+                        if (opaqueEntry.getVersion() <= persistedOpaqueEntryTs) {
+                            log.trace("Skipping entry {} as it is less than the last applied opaque entry {}",
+                                opaqueEntry.getVersion(), persistedOpaqueEntryTs);
+                            return null;
+                        }
+
+                        // LAST_OPAQUE_ENTRY_PROCESSED has the timestamp of the last OpaqueEntry applied from a
+                        // batch of opaque entries received.
+                        logReplicationMetadataManager.appendUpdate(txnContext,
+                            LogReplicationMetadataType.LAST_OPAQUE_ENTRY_PROCESSED, opaqueEntry.getVersion());
+
+                        // CorfuStore uses WriteAfterWriteTransaction type so even though the topology is read
+                        // and validated for each OpaqueEntry, this read will not be used to detect a concurrent
+                        // topology update on the Sink.  So force an update to this key by using the touch() api.
+                        // NOTE: This will be addressed once the schema of the metadata table is updated to have
+                        // all types in a single key (remote session)
+                        logReplicationMetadataManager.touch(txnContext, LogReplicationMetadataType.TOPOLOGY_CONFIG_ID);
+
+                        // If this is the last OpaqueEntry in the message/batch, update LAST_LOG_ENTRY_PROCESSED
+                        // with its timestamp
+                        if (opaqueEntry.getVersion() == txMessage.getMetadata().getTimestamp()) {
+                            logReplicationMetadataManager.appendUpdate(txnContext,
+                                LogReplicationMetadataType.LAST_LOG_ENTRY_PROCESSED, opaqueEntry.getVersion());
```

Comment:
I know they have the same value, but for reading purposes it might be better to put: txMessage.getMetadata().getTimestamp()

Reply:
agreed